File: backend/app/schemas/alert_workflow_chat.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.alert import AlertWorkflowCreate, AlertWorkflowOut
from app.schemas.system_config import LlmProvider


WorkflowChatStatus = Literal["queued", "running", "completed", "failed", "cancelled"]


class AlertWorkflowChatPreferenceOut(BaseModel):
    default_provider: LlmProvider | None = None
    default_model: str | None = None


class AlertWorkflowChatPreferenceUpdateIn(BaseModel):
    default_provider: LlmProvider | None = None
    default_model: str | None = Field(default=None, max_length=256)


class AlertWorkflowChatSessionCreateIn(BaseModel):
    title: str | None = Field(default=None, max_length=256)
    workflow_id: str | None = None
    draft_workflow: AlertWorkflowCreate | None = None


class AlertWorkflowChatSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    workflow_id: str | None = None
    title: str
    status: str = "active"
    active_snapshot_id: str | None = None
    created_at: datetime
    updated_at: datetime
    workflow: AlertWorkflowOut | None = None


class AlertWorkflowChatSubmitIn(BaseModel):
    message: str = Field(..., min_length=1)
    session_id: str | None = None
    session_title: str | None = Field(default=None, max_length=256)
    workflow_id: str | None = None
    draft_workflow: AlertWorkflowCreate | None = None
    editor_payload: dict[str, Any] = Field(default_factory=dict)
    provider: LlmProvider | None = None
    model: str | None = Field(default=None, max_length=256)
    metadata: dict[str, Any] = Field(default_factory=dict)


class AlertWorkflowChatRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    session_id: str
    user_id: str
    workflow_id: str | None = None
    status: WorkflowChatStatus | str
    job_id: str | None = None
    provider: str
    model_id: str
    message: str
    response_text: str = ""
    error: str | None = None
    metadata_json: str = "{}"
    queued_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AlertWorkflowChatSubmitOut(BaseModel):
    run: AlertWorkflowChatRunOut
    session: AlertWorkflowChatSessionOut
    stream_url: str
    status_url: str
    events_url: str


class AlertWorkflowChatEventOut(BaseModel):
    id: str
    run_id: str
    sequence: int
    event_type: str
    payload: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class AlertWorkflowChatEventsPageOut(BaseModel):
    run: AlertWorkflowChatRunOut
    events: list[AlertWorkflowChatEventOut] = Field(default_factory=list)
    next_after_sequence: int | None = None


class AlertWorkflowChatSnapshotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    session_id: str
    run_id: str | None = None
    workflow_id: str
    user_id: str
    version: int
    label: str
    workflow_payload: dict[str, Any] = Field(default_factory=dict)
    validation: dict[str, Any] = Field(default_factory=dict)
    compile: dict[str, Any] = Field(default_factory=dict)
    explanation: dict[str, Any] = Field(default_factory=dict)
    samples: dict[str, Any] = Field(default_factory=dict)
    diff: dict[str, Any] = Field(default_factory=dict)
    valid: bool
    applied_at: datetime | None = None
    created_at: datetime


class AlertWorkflowChatSnapshotApplyOut(BaseModel):
    snapshot: AlertWorkflowChatSnapshotOut
    workflow: AlertWorkflowOut

File: frontend/service/types/alert-workflow-chat.ts
import type { AlertWorkflow, LlmProvider } from "@/service/types/alerts";

export type WorkflowChatStatus = "queued" | "running" | "completed" | "failed" | "cancelled" | string;
export type WorkflowChatPayload = Record<string, unknown>;

export interface AlertWorkflowChatPreference {
    default_provider?: LlmProvider | null;
    default_model?: string | null;
}

export interface AlertWorkflowChatPreferenceUpdate {
    default_provider?: LlmProvider | null;
    default_model?: string | null;
}

export interface AlertWorkflowChatSession {
    id: string;
    user_id: string;
    workflow_id?: string | null;
    title: string;
    status: string;
    active_snapshot_id?: string | null;
    created_at: string;
    updated_at: string;
    workflow?: AlertWorkflow | null;
}

export interface AlertWorkflowChatRun {
    id: string;
    session_id: string;
    user_id: string;
    workflow_id?: string | null;
    status: WorkflowChatStatus;
    job_id?: string | null;
    provider: string;
    model_id: string;
    message: string;
    response_text: string;
    error?: string | null;
    metadata_json: string;
    queued_at: string;
    started_at?: string | null;
    completed_at?: string | null;
    created_at: string;
    updated_at: string;
}

export interface AlertWorkflowChatEvent {
    id: string;
    run_id: string;
    sequence: number;
    event_type: string;
    payload: WorkflowChatPayload;
    created_at: string;
}

export interface AlertWorkflowChatEventsPage {
    run: AlertWorkflowChatRun;
    events: AlertWorkflowChatEvent[];
    next_after_sequence?: number | null;
}

export interface AlertWorkflowChatSnapshot {
    id: string;
    session_id: string;
    run_id?: string | null;
    workflow_id: string;
    user_id: string;
    version: number;
    label: string;
    workflow_payload: WorkflowChatPayload;
    validation: WorkflowChatPayload;
    compile: WorkflowChatPayload;
    explanation: WorkflowChatPayload;
    samples: WorkflowChatPayload;
    diff: WorkflowChatPayload;
    valid: boolean;
    applied_at?: string | null;
    created_at: string;
}

export interface AlertWorkflowChatSubmitRequest {
    message: string;
    session_id?: string | null;
    session_title?: string | null;
    workflow_id?: string | null;
    draft_workflow?: WorkflowChatPayload | null;
    editor_payload?: WorkflowChatPayload;
    provider?: LlmProvider | null;
    model?: string | null;
    metadata?: WorkflowChatPayload;
}

export interface AlertWorkflowChatSubmitResponse {
    run: AlertWorkflowChatRun;
    session: AlertWorkflowChatSession;
    stream_url: string;
    status_url: string;
    events_url: string;
}

export interface AlertWorkflowChatSnapshotApplyResult {
    snapshot: AlertWorkflowChatSnapshot;
    workflow: AlertWorkflow;
}

export interface AlertWorkflowChatQueueHealth {
    queue_name: string;
    base_queue_name: string;
    queue_fingerprint: string;
    queued_count: number;
    oldest_job_id?: string | null;
    oldest_queued_seconds?: number | null;
    workers: Array<Record<string, unknown>>;
    active_worker_count: number;
    external_worker_count: number;
    fallback_worker_count: number;
    fallback_worker_available: boolean;
    has_active_worker: boolean;
    in_process_worker_enabled: boolean;
    has_processing_path: boolean;
}
